### ISSUE TITLE
ability to override default location block to support single-page-applications

### ISSUE THREAD

**Issue by srang** [sclorg/nginx-container#58]
Right now the nginx.conf provided by s2i by default includes a block:
```
 location {
 }
```
in the `default_server` block. When serving an SPA at root this causes problems. There are a bunch of examples out there for how to serve single page apps from root 

* https://gist.github.com/dimitardanailov/7a7c4e3be9e03d1b578a
* https://gkedge.gitbooks.io/react-router-in-the-real/content/nginx.html
* https://serverfault.com/questions/773705/make-nginx-serve-index-html-if-requested-url-is-not-found-without-using-error-pa

What I'd like to propose (and will submit a PR for) is to allow an environment flag for replacing the default location block with SPA block.. or if we want to add the ability to inject a SPA block through a predefined named file.

**Comment by omron93** [comment]
> Right now the nginx.conf provided by s2i by default includes a block:
>
> location {
> }

@srang I can see empty `location` block only in test app or examples. What file do you think?

**Comment by srang** [comment]
I was gonna put up a PR with an update to assemble that looks for a file called ‘nginx-default-loc.conf’ and if found, sed’s the default location block with an include for that new file

**Comment by omron93** [comment]
To be honest I'm not sure what you want to do - from PR it'll be more clear I think.

So guessing: following can't be used?

**./nginx-default-cfg/*.conf**
Contains any nginx config snippets to include in the default server block

**Comment by srang** [comment]
The issue I ran into when trying to serve an SPA using ‘nginx-default-cfg/‘ was that when I needed to serve the app from root it conflicted with the existing location block. I’ll get up the PR and we can review and figure out if it makes sense (I can also throw together a gist example of what I’m trying to allow for)

**Comment by omron93** [comment]
Thanks. I hope I finally understand. Sorry.

It makes sense, but I'm not nginx expert. @notroj What do you think?

**Comment by jkroepke** [comment]
/cc @atamanroman

Any updates here? We are hitting this bug. We want to use this image not for S2I builds just for plain docker build.

```
COPY nginx.conf /opt/app-root/etc/nginx.default.d/default.conf
```

is not good enough since we need to use our own 

```
location / {
  try_files $uri $uri/ /index.html;
}
```

Nginx does not allow to define a location twice and I don't see the reason why there is a empty `location /`.

**Comment by srang** [comment]
@jkroepke in the short term you might be able to pull out the `sed` snippet from above and then run your copy command into `nginx.default.d`.

`sed` snippet for reference: 
```
sed -i '/^\s*location \/ {/ { N; /^\s*location \/ {\n\s*}/d }' "${NGINX_CONF_PATH}"
```

I did have some permission problems with running the inplace (`-i`) sed on the base nginx.conf (hence why I didn't actually do inplace) but with a docker build you might be able to get away with it

**Comment by jkroepke** [comment]
@srang Yeah Thank for the snippet. We steal the code already from here ;-)

https://github.com/srang/angular-5-example/blob/rhel/nginx/.s2i/bin/assemble#L10

But this shouldn't be the workaround in 2018.

**Comment by srang** [comment]
@jkroepke yeah it'd be great if it was a little easier

soooo @omron93 @notroj do we think this is a good feature candidate? If so I can finalize a PR with tests

**Comment by commandtab** [comment]
Yes please! I've run into the same issue while deploying single-page apps. My current solution to avoiding conflicting `location /` blocks is to instead specify a whole new `server` in `nginx-cfg/spa.conf` like so:

```
server {
    listen 8081;
    server_name app;
    root /opt/app-root/src;
    location / {
        try_files $uri $uri/ /index.html;
    }
}
```

But then I'm wedded to `/opt/app-root/src`, which is kind of an implementation detail, but more importantly, I have to run the new `server` on port 8081 to avoid colliding with the default server over on 8080. And because I'm running on a non-default port, I have to modify my Service, DeploymentConfig, and Route resources accordingly. This works well, but it's a lot to have to remember to set up for each app, and to recall the details of when doing maintenance.

It would be great if I didn't have to inherit so many custom configurations when all I really want is `try_files $uri $uri/ /index.html;` inside the default `location` block.

Thanks! ❤️

**Comment by lholmquist** [comment]
This is the generated nginx.conf from my running container,  https://gist.github.com/lholmquist/d9c3138b1253f4f6ef72e6eaef2f9a5b

I just modified this where i needed to, then supplied it to the image, so it would use this one.  a bit of a hack, but i think it gets around the issue for now

**Comment by misanche** [comment]
Hi I think that to solve this issue the best way could be adding the following:
`
        location / {
           include <otherPath>;
        }
`

**Comment by InfoSec812** [comment]
I **JUST** hit this same issue and it's still open from 2 years ago. @jkroepke Any updates?

**Comment by jkroepke** [comment]
> @jkroepke Any updates?

Yes. I use the sed from here (https://github.com/sclorg/nginx-container/issues/58#issuecomment-400435116) in my downstream images.
